File: simulation.go
package main

// SimulateRun performs one iteration of the Monte Carlo simulation.
// It calculates the collisions, applies the MPR (Multi-Packet Reception) rule,
// and returns the successful transmission indices for each sequence.
func SimulateRun(shiftedSgSet [][]int, r int, pq int) [][]int {
	se := len(shiftedSgSet)

	// 1. Count occurrences of each slot (equivalent to addSg in MATLAB)
	slotCounts := make([]int, pq)
	for _, seqIndices := range shiftedSgSet {
		for _, idx := range seqIndices {
			slotCounts[idx]++
		}
	}

	// 2. Identify valid slots where 0 < count <= r (equivalent to rboth in MATLAB)
	validSlots := make([]bool, pq)
	for idx, count := range slotCounts {
		if count > 0 && count <= r {
			validSlots[idx] = true
		}
	}

	// 3. Find successful transmissions for each user (equivalent to intersect in MATLAB)
	// Since we only need to check if a user's transmission slot is valid,
	// we just filter their seqIndices array.
	sgSuccessCounts := make([][]int, se)
	for i, seqIndices := range shiftedSgSet {
		// Pre-allocate to optimize memory (capacity = len(seqIndices), length = 0)
		success := make([]int, 0, len(seqIndices))
		for _, idx := range seqIndices {
			if validSlots[idx] {
				success = append(success, idx)
			}
		}
		sgSuccessCounts[i] = success
	}

	return sgSuccessCounts
}

File: python_version/simulation.py
import numpy as np

def simulate_run(shifted_matrix, r, pq):
    """
    Simulates collisions and applies MPR rule.
    Returns a list of 1D numpy arrays containing sorted successful indices for each user.
    """
    # 1. Count occurrences of each slot
    counts = np.bincount(shifted_matrix.ravel(), minlength=pq)
    
    # 2. Identify valid slots (where count > 0 and <= r)
    valid_slots = (counts > 0) & (counts <= r)
    
    # 3. Find successful transmissions for each user
    # valid_slots[shifted_matrix] gives a boolean mask of shape (se, w)
    mask = valid_slots[shifted_matrix]
    
    se = shifted_matrix.shape[0]
    success_counts = []
    
    # Extract the successful indices for each user and sort them
    for i in range(se):
        success_indices = shifted_matrix[i][mask[i]]
        success_indices.sort() # In-place sort, necessary for Delay/AoI calculation
        success_counts.append(success_indices)
        
    return success_counts
